bee-crypto/src/ternary/sponge/batched_curlp/mod.rs

```diff
@@ -0,0 +1,197 @@
+
+use crate::ternary::sponge::{CurlP, CurlPRounds, Sponge};
+use bct::BCTritBuf;
+use bct_curlp::BCTCurlP;
+
+/// The number of inputs that can be processed in a single batch.
+pub const BATCH_SIZE: usize = 8 * std::mem::size_of::<usize>();
+const HIGH_BITS: usize = usize::max_value();
+/// A hasher that can process several inputs at the same time in batches.
+///
+/// This hasher works by interleaving the trits of the inputs in each batch and hashing this
+/// interleaved representation. It is also able to fall back to the regular CurlP algorithm if
+/// required.
+pub struct BatchHasher {
+    /// The trits of the inputs before being interleaved.
+    trit_inputs: Vec<TritBuf>,
+    /// An interleaved representation of the input trits.
+    bct_inputs: BCTritBuf,
+    /// An interleaved representation of the output trits.
+    bct_hashes: BCTritBuf,
+    /// The CurlP hasher for binary coded trits.
+    bct_curlp: BCTCurlP,
+    /// The regular CurlP hasher.
+    curlp: CurlP,
+}
+
+impl BatchHasher {
+    /// Create a new hasher.
+    ///
+    /// It requires the length of the input, the length of the output hash and the number of
+    /// rounds.
+    pub fn new(input_length: usize, hash_length: usize, rounds: CurlPRounds) -> Self {
+        Self {
+            trit_inputs: Vec::with_capacity(BATCH_SIZE),
+            bct_inputs: BCTritBuf::zeros(input_length),
+            bct_hashes: BCTritBuf::zeros(hash_length),
+            bct_curlp: BCTCurlP::new(hash_length, rounds as usize),
+            curlp: CurlP::new(rounds),
+        }
+    }
+    /// Add a new input to the batch.
+    ///
+    /// It panics if the size of the batch exceeds `BATCH_SIZE`.
+    pub fn add(&mut self, input: TritBuf) {
+        assert!(self.trit_inputs.len() <= BATCH_SIZE);
+        self.trit_inputs.push(input);
+    }
+    /// Return the length of the current batch.
+    pub fn len(&self) -> usize {
+        self.trit_inputs.len()
+    }
+    /// Multiplex or interleave the input trits in the bash.
+    ///
+    /// Before doing the actual interleaving, each trit is encoded as two bits which are usually
+    /// refered as the low and high bit.
+    ///
+    /// | Trit | Low bit | High bit |
+    /// |------|---------|----------|
+    /// |  -1  |    1    |     0    |
+    /// |   0  |    0    |     1    |
+    /// |   1  |    1    |     1    |
+    ///
+    /// Then the low and high bits are interleaved into two vectors of integers. Each integer has
+    /// size `BATCH_SIZE` and there are `input_length` integers in each vector.  This means that
+    /// the low and high bits of the transaction number `N` in the batch are stored in the position
+    /// `N` of each integer.
+    ///
+    /// This step works correctly even if there are less than `BATCH_SIZE` inputs.
+    fn mux(&mut self) {
+        let count = self.trit_inputs.len();
+        for i in 0..self.bct_inputs.len() {
+            let bc_trit = self.bct_inputs.get_mut(i);
```

Comment:
Since the index comes from iteration on itself, I would use an unsafe version of `get_mut`, that assert is definitely not needed